File: algorithms/1694.reformat-phone-number/solution.rs
impl Solution {
    pub fn reformat_number(number: String) -> String {
        use std::collections::VecDeque;

        let mut parts: Vec<String> = vec![];
        let mut nums: VecDeque<char> = number.clone().replace("-", "").replace(" ", "").chars().collect();

        let mut three_digits = nums.len() / 3;
        let mut two_digits = match nums.len() % 3 {
            1 => {
                three_digits -= 1;
                2
            },
            2 => 1,
            _ => 0,
        };

        for _ in 0..three_digits {
            let mut part: String = "".to_string();
            for _ in 0..3 {
                part.push_str(&(nums.pop_front().unwrap().to_string()));
            }
            parts.push(part);
        }

        for _ in 0..two_digits {
            let mut part: String = "".to_string();
            for _ in 0..2 {
                part.push_str(&(nums.pop_front().unwrap().to_string()));
            }
            parts.push(part);
        }

        parts.join("-")
    }
})

File: algorithms/1694.reformat-phone-number/solution.go
func reformatNumber(number string) string {
	parts := []string{}
	nums := []rune{}

	for _, r := range number {
		if r != '-' && r != ' ' {
			nums = append(nums, r)
		}
	}

	threeDigits := len(nums) / 3
	twoDigits := 0
	switch len(nums) % 3 {
	case 1:
		threeDigits -= 1
		twoDigits = 2
	case 2:
		twoDigits = 1
	default:
		twoDigits = 0
	}

	for i := 0; i < threeDigits; i++ {
		var s string

		s += string(nums[0:3])
		nums = nums[3:]

		parts = append(parts, s)
	}

	for i := 0; i < twoDigits; i++ {
		var s string

		s += string(nums[0:2])
		nums = nums[2:]

		parts = append(parts, s)
	}

	return strings.Join(parts, "-")
}

